presto-product-tests/src/main/java/io/prestosql/tests/hive/TestHiveTransactionalTable.java

```diff
@@ -184,19 +183,58 @@ private static String hiveTableProperties(TransactionalTableType transactionalTa
 
     private static void compactTableAndWait(CompactionMode compactMode, String tableName, String partitionString, Duration timeout)
     {
-        assertEquals(getTableCompactions(compactMode, tableName).count(), 0);
+        Failsafe.with(
+                new RetryPolicy<>()
+                        .withMaxDuration(java.time.Duration.ofMillis(timeout.toMillis()))
+                        .withMaxAttempts(Integer.MAX_VALUE))  // limited by MaxDuration
+                .onFailure(event -> {
+                    throw new IllegalStateException("Could not compact table, retries " + event.getAttemptCount(), event.getFailure());
+                })
+                .run(() -> tryCompactingTable(compactMode, tableName, partitionString, timeout));
+    }
+
+    private static void tryCompactingTable(CompactionMode compactMode, String tableName, String partitionString, Duration timeout)
+            throws TimeoutException
+    {
+        // Hive's unix_timestamp() returns millis
+        long beforeCompactionStart = (Long) onHive().executeQuery("SELECT unix_timestamp()").row(0).get(0) * 1000;
         onHive().executeQuery(format("ALTER TABLE %s %s COMPACT '%s'", tableName, partitionString, compactMode.name())).getRowsCount();
 
-        // Since we disabled table auto compaction and we checked that there are no compaction to the table
-        // we can assume that every compaction from now on is triggered in this test
-        // and all compaction should complete successfully before proceeding.
-        assertEventually(timeout, () -> {
-            Map<String, String> compaction = getOnlyElement(getTableCompactions(compactMode, tableName)
-                    .collect(toImmutableList()));
+        long loopStart = System.nanoTime();
+        while (true) {
+            try {
+                // Compaction takes couple of second so there is no need to check state more frequent than 1s
+                Thread.sleep(1000);
+            }
+            catch (InterruptedException e) {
+                throw new RuntimeException(e);
+            }
+
+            // Since we disabled auto compaction for uniquely named table and every compaction is triggered in this test
+            // we can expect that single compaction in given mode should complete before proceeding.
+            Optional<Map<String, String>> startedCompaction = getTableCompactions(compactMode, tableName)
+                    .filter(row -> !row.get("state").equals("initiated"))
+                    .filter(row -> Long.parseLong(row.get("start time")) >= beforeCompactionStart)
+                    .findFirst();
```

Comment:
should we take all and then assert that the size of list is either 0 or 1

Reply:
We are filtering compaction that started after `beforeCompactionStart` so there should be exactly [0;1] compaction in state {running,failed,completed,cleaning up}.

Reply:
Yeah - I know. Just it would be nice to assert that. Not blindly take first.